Continuous Delivery not working with OpenSearch

PDS Jenkins had been happily delivering the `cd` branch of the `registry` repository for some time. However, that branch was using Elasticsearch for the `elasticsearch` container.

After switching from `cd` → `main`, Jenkins could no longer deploy the registry service.

There was an issue with leftover directories that caused `generate-certs.sh` to fail, but even after cleaning that up by hand, the `elasticsearch` container (now running OpenSearch) crashes at startup.

The `Jenkinsfile` will need to be updated but before so we need to figure out why the Docker Composition profile `int-registry-batch-loader` fails.

Here's the stack trace from the `elasticsearch` container running OpenSearch:
```
org.opensearch.bootstrap.StartupException: java.lang.IllegalStateException: failed to load plugin class [org.opensearch.security.OpenSearchSecurityPlugin]
 at org.opensearch.bootstrap.OpenSearch.init(OpenSearch.java:186) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.bootstrap.OpenSearch.execute(OpenSearch.java:173) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:99) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.cli.Command.mainWithoutErrorHandling(Command.java:140) ~[opensearch-cli-1.1.0.jar:1.1.0]
 at org.opensearch.cli.Command.main(Command.java:103) ~[opensearch-cli-1.1.0.jar:1.1.0]
 at org.opensearch.bootstrap.OpenSearch.main(OpenSearch.java:139) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.bootstrap.OpenSearch.main(OpenSearch.java:105) ~[opensearch-1.1.0.jar:1.1.0]
Caused by: java.lang.IllegalStateException: failed to load plugin class [org.opensearch.security.OpenSearchSecurityPlugin]
 at org.opensearch.plugins.PluginsService.loadPlugin(PluginsService.java:736) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.plugins.PluginsService.loadBundle(PluginsService.java:672) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.plugins.PluginsService.loadBundles(PluginsService.java:492) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.plugins.PluginsService.<init>(PluginsService.java:181) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.node.Node.<init>(Node.java:357) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.node.Node.<init>(Node.java:301) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.bootstrap.Bootstrap$5.<init>(Bootstrap.java:240) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.bootstrap.Bootstrap.setup(Bootstrap.java:240) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.bootstrap.Bootstrap.init(Bootstrap.java:406) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.bootstrap.OpenSearch.init(OpenSearch.java:182) ~[opensearch-1.1.0.jar:1.1.0]
 ... 6 more
Caused by: java.lang.reflect.InvocationTargetException
 at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:?]
 at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:64) ~[?:?]
 at jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:?]
 at java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:500) ~[?:?]
 at java.lang.reflect.Constructor.newInstance(Constructor.java:481) ~[?:?]
 at org.opensearch.plugins.PluginsService.loadPlugin(PluginsService.java:727) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.plugins.PluginsService.loadBundle(PluginsService.java:672) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.plugins.PluginsService.loadBundles(PluginsService.java:492) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.plugins.PluginsService.<init>(PluginsService.java:181) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.node.Node.<init>(Node.java:357) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.node.Node.<init>(Node.java:301) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.bootstrap.Bootstrap$5.<init>(Bootstrap.java:240) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.bootstrap.Bootstrap.setup(Bootstrap.java:240) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.bootstrap.Bootstrap.init(Bootstrap.java:406) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.bootstrap.OpenSearch.init(OpenSearch.java:182) ~[opensearch-1.1.0.jar:1.1.0]
 ... 6 more
Caused by: org.opensearch.OpenSearchSecurityException: Error while initializing transport SSL layer from PEM: OpenSearchException[Unable to read /usr/share/opensearch/config/esnode-key.pem (/usr/share/opensearch/config/esnode-key.pem). Please make sure this files exists and is readable regarding to permissions. Property: plugins.security.ssl.transport.pemkey_filepath]
 at org.opensearch.security.ssl.DefaultSecurityKeyStore.initTransportSSLConfig(DefaultSecurityKeyStore.java:419) ~[?:?]
 at org.opensearch.security.ssl.DefaultSecurityKeyStore.initSSLConfig(DefaultSecurityKeyStore.java:258) ~[?:?]
 at org.opensearch.security.ssl.DefaultSecurityKeyStore.<init>(DefaultSecurityKeyStore.java:179) ~[?:?]
 at org.opensearch.security.ssl.OpenSearchSecuritySSLPlugin.<init>(OpenSearchSecuritySSLPlugin.java:218) ~[?:?]
 at org.opensearch.security.OpenSearchSecurityPlugin.<init>(OpenSearchSecurityPlugin.java:252) ~[?:?]
 at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:?]
 at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:64) ~[?:?]
 at jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:?]
 at java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:500) ~[?:?]
 at java.lang.reflect.Constructor.newInstance(Constructor.java:481) ~[?:?]
 at org.opensearch.plugins.PluginsService.loadPlugin(PluginsService.java:727) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.plugins.PluginsService.loadBundle(PluginsService.java:672) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.plugins.PluginsService.loadBundles(PluginsService.java:492) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.plugins.PluginsService.<init>(PluginsService.java:181) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.node.Node.<init>(Node.java:357) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.node.Node.<init>(Node.java:301) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.bootstrap.Bootstrap$5.<init>(Bootstrap.java:240) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.bootstrap.Bootstrap.setup(Bootstrap.java:240) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.bootstrap.Bootstrap.init(Bootstrap.java:406) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.bootstrap.OpenSearch.init(OpenSearch.java:182) ~[opensearch-1.1.0.jar:1.1.0]
 ... 6 more
Caused by: org.opensearch.OpenSearchException: Unable to read /usr/share/opensearch/config/esnode-key.pem (/usr/share/opensearch/config/esnode-key.pem). Please make sure this files exists and is readable regarding to permissions. Property: plugins.security.ssl.transport.pemkey_filepath
 at org.opensearch.security.ssl.DefaultSecurityKeyStore.checkPath(DefaultSecurityKeyStore.java:952) ~[?:?]
 at org.opensearch.security.ssl.DefaultSecurityKeyStore.resolve(DefaultSecurityKeyStore.java:237) ~[?:?]
 at org.opensearch.security.ssl.DefaultSecurityKeyStore.initTransportSSLConfig(DefaultSecurityKeyStore.java:397) ~[?:?]
 at org.opensearch.security.ssl.DefaultSecurityKeyStore.initSSLConfig(DefaultSecurityKeyStore.java:258) ~[?:?]
 at org.opensearch.security.ssl.DefaultSecurityKeyStore.<init>(DefaultSecurityKeyStore.java:179) ~[?:?]
 at org.opensearch.security.ssl.OpenSearchSecuritySSLPlugin.<init>(OpenSearchSecuritySSLPlugin.java:218) ~[?:?]
 at org.opensearch.security.OpenSearchSecurityPlugin.<init>(OpenSearchSecurityPlugin.java:252) ~[?:?]
 at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:?]
 at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:64) ~[?:?]
 at jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:?]
 at java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:500) ~[?:?]
 at java.lang.reflect.Constructor.newInstance(Constructor.java:481) ~[?:?]
 at org.opensearch.plugins.PluginsService.loadPlugin(PluginsService.java:727) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.plugins.PluginsService.loadBundle(PluginsService.java:672) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.plugins.PluginsService.loadBundles(PluginsService.java:492) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.plugins.PluginsService.<init>(PluginsService.java:181) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.node.Node.<init>(Node.java:357) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.node.Node.<init>(Node.java:301) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.bootstrap.Bootstrap$5.<init>(Bootstrap.java:240) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.bootstrap.Bootstrap.setup(Bootstrap.java:240) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.bootstrap.Bootstrap.init(Bootstrap.java:406) ~[opensearch-1.1.0.jar:1.1.0]
 at org.opensearch.bootstrap.OpenSearch.init(OpenSearch.java:182) ~[opensearch-1.1.0.jar:1.1.0]
 ... 6 more
uncaught exception in thread [main]
java.lang.IllegalStateException: failed to load plugin class [org.opensearch.security.OpenSearchSecurityPlugin]
Likely root cause: OpenSearchException[Unable to read /usr/share/opensearch/config/esnode-key.pem (/usr/share/opensearch/config/esnode-key.pem). Please make sure this files exists and is readable regarding to permissions. Property: plugins.security.ssl.transport.pemkey_filepath]
 at org.opensearch.security.ssl.DefaultSecurityKeyStore.checkPath(DefaultSecurityKeyStore.java:952)
 at org.opensearch.security.ssl.DefaultSecurityKeyStore.resolve(DefaultSecurityKeyStore.java:237)
 at org.opensearch.security.ssl.DefaultSecurityKeyStore.initTransportSSLConfig(DefaultSecurityKeyStore.java:397)
 at org.opensearch.security.ssl.DefaultSecurityKeyStore.initSSLConfig(DefaultSecurityKeyStore.java:258)
 at org.opensearch.security.ssl.DefaultSecurityKeyStore.<init>(DefaultSecurityKeyStore.java:179)
 at org.opensearch.security.ssl.OpenSearchSecuritySSLPlugin.<init>(OpenSearchSecuritySSLPlugin.java:218)
 at org.opensearch.security.OpenSearchSecurityPlugin.<init>(OpenSearchSecurityPlugin.java:252)
 at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
 at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:64)
 at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
 at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:500)
 at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:481)
 at org.opensearch.plugins.PluginsService.loadPlugin(PluginsService.java:727)
 at org.opensearch.plugins.PluginsService.loadBundle(PluginsService.java:672)
 at org.opensearch.plugins.PluginsService.loadBundles(PluginsService.java:492)
 at org.opensearch.plugins.PluginsService.<init>(PluginsService.java:181)
 at org.opensearch.node.Node.<init>(Node.java:357)
 at org.opensearch.node.Node.<init>(Node.java:301)
 at org.opensearch.bootstrap.Bootstrap$5.<init>(Bootstrap.java:240)
 at org.opensearch.bootstrap.Bootstrap.setup(Bootstrap.java:240)
 at org.opensearch.bootstrap.Bootstrap.init(Bootstrap.java:406)
 at org.opensearch.bootstrap.OpenSearch.init(OpenSearch.java:182)
 at org.opensearch.bootstrap.OpenSearch.execute(OpenSearch.java:173)
 at org.opensearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:99)
 at org.opensearch.cli.Command.mainWithoutErrorHandling(Command.java:140)
 at org.opensearch.cli.Command.main(Command.java:103)
 at org.opensearch.bootstrap.OpenSearch.main(OpenSearch.java:139)
 at org.opensearch.bootstrap.OpenSearch.main(OpenSearch.java:105)
For complete error details, refer to the log at /usr/share/opensearch/logs/docker-cluster.log
Killing performance analyzer process 35
OpenSearch exited with code 1
```

--- a/Jenkinsfile
+++ b/Jenkinsfile
@@ -93,11 +93,18 @@ pipeline {
             steps {
                 dir("${env.WORKSPACE}/docker") {
                     sh "$compose down --remove-orphans --timeout ${shutdown_timeout} --volumes ||:"
-                    // 🔮 TODO: Include --no-color? 
-                    sh "$compose up --detach --quiet-pull --timeout ${shutdown_timeout}"
+                }
+                dir("${env.WORKSPACE}/docker/certs") {
+                    // Now re-generate some certificates:
+                    sh "./generate-certs.sh"
+                }
+                dir("${env.WORKSPACE}/docker") {
+                    // Get the latest and greatest
+                    sh "$compose pull --quiet"
+                    // Make a bunch of containers and launch 'em; 🔮 TODO: Include --no-color? 
+                    sh "$compose up --detach --timeout ${shutdown_timeout}"
                 }
             }
-
             // 🔮 TODO: Include a `post {…}` block to do post-deployment test queries?
         }
     }

--- a/docker/certs/generate-certs.sh
+++ b/docker/certs/generate-certs.sh
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/bin/sh -e
 
 # ----------------------------------------------------------------------------------------------------------------
 #
@@ -8,6 +8,10 @@
 # Read more information at: https://opensearch.org/docs/latest/security-plugin/configuration/generate-certificates/
 #
 # ----------------------------------------------------------------------------------------------------------------
+
+# Sometimes—and for reasons we can't yet fathom—these get generated as _directories_ from earlier
+# runs. So let's make sure they're _gone_ before we try anything:
+rm -rf node1-key.pem node1.pem root-ca-key.pem root-ca.pem root-ca.srl
 
 # Root CA
 openssl genrsa -out root-ca-key.pem 2048
@@ -20,5 +24,6 @@ openssl req -new -key node1-key.pem -subj "/C=CA/ST=CALIFORNIA/L=LA/O=ORG/OU=PDS
 openssl x509 -req -in node1.csr -CA root-ca.pem -CAkey root-ca-key.pem -CAcreateserial -sha256 -out node1.pem -days 730
 
 # Cleanup
-rm node1-key-temp.pem
-rm node1.csr
+# Note: you normally don't want your private keys world-readable, but this makes Docker happy.
+chmod a+r node1-key.pem node1.pem root-ca-key.pem root-ca.pem root-ca.srl
+rm node1-key-temp.pem node1.csr

